_posts/2024-06-14-granite-rag.adoc

```diff
@@ -0,0 +1,278 @@
+
+The `maxResults` method in the EmbeddingStoreContentRetriever builder specifies the number of text segments to retrieve.
+Since our segments are large, we retrieve only two segments.
+
+=== The AI Service
+
+The https://github.com/cescoffier/quarkus-granite-rag-demo/blob/main/src/main/java/me/escoffier/granite/rag/ChatBot.java[AI Service (ChatBot.java)] is the core component of our chatbot, combining the capabilities of the retriever and the Granite LLM to generate appropriate responses.
+
+With Quarkus, implementing an AI service is straightforward:
+
+[source,java]
+----
+@RegisterAiService(retrievalAugmentor = Retriever.class) // <1>
+@SystemMessage("You are Mona, a chatbot answering question about a museum. Be polite, concise and helpful.") // <2>
+@SessionScoped // <3>
+public interface ChatBot {
+
+    String chat(String question); // <4>
+
+}
+----
+<1> The `@RegisterAiService` annotation specifies the retrieval augmentor to use, which in our case is the `Retriever` bean defined earlier.
+<2> The `@SystemMessage` annotation provides the main instructions for the AI model.
+<3> The `@SessionScoped` annotation ensures that the AI service is stateful, maintaining context between user interactions for more engaging conversations.
+<4> The `ChatBot` interface defines a single method, `chat`, which takes a user question as input and returns the chatbot's response.
+
+=== The WebSocket endpoint
+
+The final piece is the https://github.com/cescoffier/quarkus-granite-rag-demo/blob/main/src/main/java/me/escoffier/granite/rag/ChatWebSocket.java[WebSocket endpoint (ChatWebSocket.java)], which serves as the communication bridge between the chatbot's backend and the frontend interface:
+
+[source,java]
+----
+@WebSocket(path = "/chat") // <1>
+public class ChatWebSocket {
+
+    @Inject ChatBot bot; // Inject the AI service
+
+    @OnOpen // <2>
+    String welcome() {
+        return "Welcome, my name is Mona, how can I help you today?";
+    }
+
+    @OnTextMessage // <3>
+    String onMessage(String message) {
+        return bot.chat(message);
+    }
+
+}
+----
+<1> The `@WebSocket` annotation specifies the WebSocket path.
+<2> The `@OnOpen` method sends a welcome message when a user connects to the _WebSocket_.
+<3> The `@OnTextMessage` method processes the user's messages and returns the chatbot's responses, using the injected AI service.
+
+That's it! Our chatbot is now ready to chat with users, providing contextually relevant responses based on the RAG pattern.
+
+== Running the application
+
+Let's run the application and see our chatbot in action.
+First, clone the https://github.com/cescoffier/quarkus-granite-rag-demo/tree/main[repository] and run the following command:
+
+[source,shell]
+----
+./mvnw quarkus:dev
+----
+
+This command starts the Quarkus application in development mode.
+Ensure you have InstructLab or Podman AI Studio running to use the Granite LLM.
```

Comment:
should we mention on windows and mac better to use instructlab due to lack of GPU utilization in containers ?